Fix double-quotes handling in subscription button

wojtekn opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes https://github.com/Automattic/wp-calypso/issues/63196

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* it adds HTML encoding for double-quote characters used in Subscribe block's submit button

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

N/A

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

##### Test the fix

1. Open a New Post editor
2. Add Subscribe block
4. Use one or more double-quotes in the button
6. Save the post
7. Open post in the frontend

All double-quotes should be displayed.

Editor view:

![Screen Shot 2022-06-16 at 08 43 39](https://user-images.githubusercontent.com/727413/174008430-829f84d1-19ec-4b36-8753-6f731bf3fa08.png)

Frontend:

![Screen Shot 2022-06-16 at 08 43 32](https://user-images.githubusercontent.com/727413/174008633-53bfa9b6-643b-480b-ad2e-62a32546665b.png)

##### Regression test

We should ensure that posts that were created before the fix can still be edited, and that re-saving such post fixes double-quotes. For that, before the fix is applied, prepare a post that includes Subscribe block, with a button that contains double-quotes.  Then:

1. Edit a Post that includes such a Subscribe block
2. Confirm that block is not broken and there is no prompt for recovery
3. Make any change to the post (add another block, or add a tag), save
4. Confirm that double quotes are displayed in frontend now

wojtekn committed:
[not verified] Fix double-quotes handling for the subscription block button

When the subscription block is saved, the submitButtonText attribute
has a double-quote encoded as a Unicode character. However, the submit_button_text
shortcode attribute is not encoded, and it breaks button text when double-quotes
are used.

This PR adds HTML encoding for double-quotes in submit_button_text. HTML encoding
is used as this text is later directly used in the generated HTML code.

wojtekn committed:
[not verified] changelog

wojtekn commented:
When I worked on the fix, I checked deprecation logic (https://developer.wordpress.org/block-editor/reference-guides/block-api/block-deprecation/), and I found out that the validator doesn't catch this as a change in the block. Thus I think there is a point in using the deprecation method. If it still needs to be used, I would appreciate the feedback.

danielbachhuber reviewed (dismissed):
👍  Works well for me!

A couple of pieces to note:
1. In order for the Subscribe block to appear, Jetpack needs to be connected to WordPress.com. In my local environment, this required setting up a ngrok tunnel.
2. Because the `submitButtonText` is already properly encoded in the block's JSON attributes, no block warning appears with this change. The change solely impacts how the value is encoded in the shortcode attributes generated in the `save()` function.

<img width="1306" alt="image" src="https://user-images.githubusercontent.com/36432/174062598-2ab5afad-d978-4c9d-a4d5-7b6da11ad961.png">

An even more comprehensive fix would be to replace double quotes when the shortcode attributes are generated: https://github.com/Automattic/jetpack/blob/a99e0ccac5aeb9092b5c01d75af6846386182738/projects/plugins/jetpack/extensions/blocks/subscriptions/save.js#L137

Shortcake did this way back in the day: https://github.com/wp-shortcake/shortcake/blob/9748053a188e513d2b53126c01f1a0c7cd835a8c/js/build/shortcode-ui.js#L318

However, for the purposes of this issue, it's probably an over-optimization. I'm happy with this pull request as it is.

anomiex commented:
> An even more comprehensive fix would be to replace double quotes when the shortcode attributes are generated:
> 
> https://github.com/Automattic/jetpack/blob/a99e0ccac5aeb9092b5c01d75af6846386182738/projects/plugins/jetpack/extensions/blocks/subscriptions/save.js#L137
> 
> Shortcake did this way back in the day: https://github.com/wp-shortcake/shortcake/blob/9748053a188e513d2b53126c01f1a0c7cd835a8c/js/build/shortcode-ui.js#L318
> 
> However, for the purposes of this issue, it's probably an over-optimization.

I was about to suggest the same change. I don't think it's an over-optimization, IMO it's less surprising for everyone reading the code to do the escaping at the place where the escaping is needed rather than doing it somewhere upstream.

Doing it there will also fix the similar problem that applies to the success message field. And would also affect the placeholder field, except I don't see an ability to edit the placeholder.

P.S. The code in the shortcake link doesn't seem too useful. I'd probably do something like this instead.
```javascript
/**
 * Apply HTML encoding to characters that are special inside shortcode attribute values.
 *
 * Shortcode attribute values may not contain brackets, angle brackets (unless valid HTML), or quotes.
 * Also non-breaking space and zero-width space will be mangled, and ampersand might otherwise be
 * interpreted as part of an entity.
 *
 * @see https://codex.wordpress.org/Shortcode_API#Attributes
 * @param {string} value - Value to encode.
 * @returns {string} Encoded value.
 */
function encodeValueForShortcodeAttribute( value ) {
	const entities = { '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;' }; 
	return value.replace( /[\[\]"'&<>\u00a0\u200b]/g, m => specials[m] || `&#${ m.codePointAt( 0 ) };` );
}
```
The `specials` bit isn't strictly necessary, but will probably be more readable if a human looks at the shortcode. Or you could put the other 5 characters into `specials` too (and rename it `entities` or something) and skip the `` || `&#${ m.codePointAt( 0 ) };` `` bit.

danielbachhuber commented:
> P.S. The code in the shortcake link doesn't seem too useful. I'd probably do something like this instead.

Oh, that is a much better example than the Shortcake link 😄 Thanks for sharing it!

wojtekn committed:
[not verified] Optimize location for double-quotes escaping

Code that handles special characters escaping is extracted
to the separate function. Then, escaping for additonal
characters is added.

wojtekn commented:
@danielbachhuber @anomiex, thank you for the feedback. I've extracted the code to the separate function and moved it to a place closer to the part that returns the value. Then I've added support for escaping for more characters. 

<img width="791" alt="Screen Shot 2022-06-20 at 13 06 01" src="https://user-images.githubusercontent.com/727413/174591879-1b366ed1-89cd-4a60-81d5-341e31b85cea.png">

wojtekn committed:
[not verified] Regenerate fixtures for the test that includes single-quotes

anomiex commented on projects/plugins/jetpack/extensions/blocks/subscriptions/save.js:
```diff
@@ -128,13 +128,31 @@ export default function Save( { className, attributes } ) {
+			.replace( /</g, '&lt;' )
+			.replace( />/g, '&gt;' )
+			.replace( /"/g, '&quot;' )
+			.replace( /'/g, '&#039;' );
```
It breaks in some cases unless we also include `[` and `]`.
```suggestion
			.replace( /'/g, '&#039;' )
			.replace( /\[/g, '&#091;' )
			.replace( /\]/g, '&#093;' );
```
And as mentioned earlier, we may want to include NBSP and ZWSP to avoid them being mangled.
```suggestion
			.replace( /'/g, '&#039;' )
			.replace( /\[/g, '&#091;' )
			.replace( /\]/g, '&#093;' )
			.replace( /\u00a0/g, '&nbsp;' )
			.replace( /\u200b/g, '&#x200b;' );
```
Looks like whether `&` should be encoded or not is tricky, as the value passed in for `submitButtonText` already has ampersands (and `<`) escaped while the value for `successMessage` does not. Hooray for consistency? I guess we should leave it out.

wojtekn replied:
@anomiex, thanks for finding those missing cases. I've added square brackets, nbsp, and zwsp to the chain. It looks like the square brackets were the biggest offender here, breaking the post entirely.

wojtekn committed:
[not verified] Add escaping for square brackets, nbsp, and zwsp

wojtekn committed:
[not verified] Remove explicit list of escaped characters from the comment.

zaguiini commented on projects/plugins/jetpack/extensions/blocks/subscriptions/save.js:
```diff
@@ -128,13 +128,35 @@ export default function Save( { className, attributes } ) {
+	 * @param {string} value - Value to encode.
+	 * @returns {string} Encoded value.
+	 */
+	const encodeValueForShortcodeAttribute = value => {
```
I think we can extract this function to some kind of utility file because here it's going to get redefined on every render, and that's kind of useless.

If you don't feel like putting it in a separate file, moving out of the scope of the component should be enough.

wojtekn replied:
Good idea, thanks. I've extracted the function to the `utils` file, which is a practice followed by other blocks.

zaguiini reviewed (dismissed):
Looking good. Left a minor, non-blocking comment.

wojtekn committed:
[not verified] Extract encoding function to the separate file

Extract the function to the separate utlity file to avoid
redefining it on every render.

wojtekn commented:
It's r247994-wpcom
